hodie.1
.\"Process this file with
.\"groff -man -Tascii hodie.1
.\"
.TH HODIE 1 "AUGUST 2000" Linux "User Manuals"
.SH NAME
hodie \- Print current date and time... in Latin
.SH SYNOPSIS
.BR "hodie " "["
.I "OPTION"
]...
.SH DESCRIPTION
.B hodie
prints out the current date using classic Latin, and in
addition also prints it out and time using Roman numerals.
.SH OPTIONS
.IP "-h, --help"
Print short help message with syntax
.IP "-v, --verbose"
Print months and days (pridie, Kalends, Nones, Ides) full and not the
respective abbreviations (standard mode of operation)

.IR "Two occurrences of " "-v" " as well as the use of " "-vv" 
.IR " or " "--extremely-verbose "
will include the numerals where applicable fully declined, as in
.RB "'ante diem " quintum " Kalends Septembres'."
.IP "-n, --numerals"
Don't print anything in Latin \- only the date and time as Roman numerals.
.IP "-x, --force-numerals"
Print both the verbose latin and the date and time as Roman numerals.
.IP "-c, --classic, --auc"
Print the year in the classic manner
.I "ab urbe condita"
instead of the more modern
.I "anno domini".
.IP "-a, --ante-diem"
Print the date expressing the number of days to the next main day with
the
.I "ante diem"
expression instead of ablative case.
.IP "-d, --date"
Print out any date. This has a rather special syntax, with a keyword
following the 
.I -d
flag choosing input format. See section on 
.I DATE INPUT
below.
.IP "-r, --republican OFFSET"
Print out the date dated 
.I ab urbe tua condita
with the offset counted in years as compared to the modern european
kalendar (originating with the hypothetical birth of christ).
.BI "hodie " "-r -753"
would be equivalent with
.BI "hodie " "-c"
.
.IP "--version"
Print out the version number of this release and exit. No matter
whether other options appear on the command line or not.
.SH DATE INPUT
Following the 
.I -d
or the
.I --date
option flags, the first item 
.B must
be one of the following:
.IP verbose
In this case, the year, month and day are given by following the
.I verbose
keyword by the flags 
.I -y, --year, -m, --month, -d, --day
for year, month and date respectively
.IP ymd
After this flag, the date comes in the format
.I YYYY-MM-DD
, where the numbers may 
be separated by any non-numeric character.
.IP dmy
With this flag, the date is given as
.I DD-MM-YYYY
.IP mdy
With this flag, the date is given as
.I MM-DD-YYYY
.
Restrictions on the characters that may replace the hyphen apply as above.
.SH HISTORY
The story began on the 10. of August, 2000 (a.d. VI Id. Iul.,
MM). Having finished most of my assignment for my two-month summer job
at Ericsson Eurolab Deutschland, Nuremberg, I was idling around on the
Internet, and stumbled over the dotcomma-challenges
.I <http://www.dotcomma.org>
, where especially the Roman numeral challenge started my mind.

Almost an hour hacking, and there it was, another hour, and the
.I "language support"
was there. Before the night was over, I had written this man page and
had the layout of a decent 
.B Makefile
drawn out mentally.

At the end of the next day, I was so far that I actually had the
workings of 
.I RPM
worked out, constructed a .rpm-package and a .src.rpm-package, which was
promptly released on my home-page, announced on freshmeat and uploaded
to metalab (apps/misc :-).

Response was quick and plentiful. By now, I have compilation reports
from Linux, FreeBSD and SCO Unixware 7; there are a few compability
issues to put aside, but it works surprisingly well.
.SH RETURN VALUES
.B hodie
returns zero. Always. If it doesn't, then something is
.B really
bad with the code.

For some really unreadable code, this means that
.B hodie
could be used as a strange replacement for
.B true
.
.SH BUGS
It doesn't sanity check the input... telling
.B hodie
to display the roman form of the 99th of march gives a slightly
jumbled output, which most definitely does not make sense.

Reports are more than welcome (e-mail below).
.SH AUTHOR
Now, who would come up with such a thing?
Well, I'm Mikael Johansson, a rather all-round geek from Stockholm. I'm
gravely interested in languages, in computers and in mathematics; a
combination more dangerous than you might think.

E-mails to
.I <mikael@johanssons.org>
.SH SEE ALSO
.BR date (1)


CHANGELOG
0.4
	Was capable of converting the date and printing it verbosely. No
	command line options accepted.

0.7
	Had a man-page, certain command line options and a brief
	error/usage message.
0.8
	Implemented the ab urbe condita mode of dates and a version
	option. 
1.0pre
	Tried building a RPM-package of it.
1.0
	Grammar checked. Due to grammar details, internal representation was
	slightly changed (had to cover both ablative and accusative in the
	verbose output).  
1.1 
	Verbosified ab urbe condita display as well. Bug fixes
	arriving on first day online fixed (platform-specific coding
	and bad month calculations).  
1.2 
	Modularised code structure... Now it is organised after one
	function, one source file, and the makefile and headers have
	been hacked to support this. 

	Man page partially rewritten. Tnx to Ted Rolle for his
	extensive tips on improvements. 

	__isleap portability issues resolved. I simply replaced all
	calls to it with calls to my own leap-year calculator.

	The  --version call is more extensive and even contains a few
	words about licensing and author :-)

	The republican switch is added, as well as a BUGS and a
	CREDITS file to the tarball.

	The month handling is slightly redone, removing all those
	nasty bugs I found as soon as I was able to play around with
	the dates on my own :-) 
1.3
	Modularized the help-texts and error messages. At least a
	bit. Those of you interested in translating the english bits
	of hodie into other languages will find most of what you need
	in strings.h.en and the manual in hodie.1.en. I will include a
	diff from the last release of these two files to allow for
	simple incremental translation. Kudos to Jose Ivan Juanes for
	doing a translation into spanish for me! 

	Fixed the core-dumping bugs. I think. Mail me if it still
	screws up. 
1.4pre
	New feature: extremely verbose display. Declines the number of
	days till the next main day as well instead of just displaying
	it as a roman numeral.

	The sourcefile split of version 1.3 has transceded into split
	source branches instead. It's easier to maintain that way...

	sscanf changed to atoi in several places in the
	parse_date.c-file -- improved security (atoi actually checks
	data and does other neat stuff :-)
	Support for any separator in date parsing added (i.e. it's
	done by a 'sscanf(str,"%d%c%d%c%d"...)' instead of a
	'sscanf(str,"%d-%d-%d"...)'. 

	Display of numerals off as default - -x turns it on.

1.4
	'extremely verbose display' now _implemented_. The function
	taking care of the declensions will be basis for another,
	parallel program to hodie -- that simply will put out any
	number in any of the wanted forms, declined as asked for.

	Support for this kind of program lies in the to_roman.c file,
	but the actual program is not yet implemented. (Due to large
	parts of my audience in this question, a Windoze port of hodie
	is coming up :-)

1.5
        Replaced Makefile with autoconf/automake

        Replaced option parsing with getopt


hodie.c
#include "includes.h"
#include <getopt.h>

const char* AUC[3]={"ab urbe condita","AUC","ab urbe condita"};

int
main(int argc, char** argv)
{
    char sec[16],min[16],hrs[16],day[16],mon[16],year[16],verbose[80],datetype[16],dateval[16];
    struct tm* p_ts;
    time_t t;
    int auc_base;
    short numerals, auc, leap, custom, ad;
    mode m;
    verb_struct verbosity;
    int verbosity_level = 0;

    int c;
    int xverbosity = 0;

    verbosity.year=0;
    verbosity.day=0;
    numerals=1;
    auc=0;
    leap=0;
    custom=0;
    auc_base=0;
    ad=0;

    t=time(NULL);
    p_ts=localtime(&t);

    /* Parsing command line arguments with getopt */

    while (1) 
    {
        int option_index = 0;
        static struct option long_options[] = {
            {"help", 0, 0, 'h'},
            {"verbose", 0, 0, 'v'},
            {"extremely-verbose", 0, 0, 'X'},
            {"version", 0, 0, 'V'},
            {"numerals", 0, 0, 'n'},
            {"force-numerals", 0, 0, 'x'},
            {"classic", 0, 0, 'c'},
            {"auc", 0, 0, 'c'},
            {"ante-diem", 0, 0, 'a'},
            {"date", 1, 0, 'd'},
            {"republican", 0, 0, 'r'},
            {"year", 0, 0, 'Y'},
            {"month", 0, 0, 'M'},
            {"day", 0, 0, 'D'},
            {0,0,0,0},
        };
        
        c = getopt_long(argc, argv, 
                        "a:cd:hr:nvxD:M:VXY:",
                        long_options, &option_index);

        if(c == -1)
            break;

        switch(c)
        {
        case 'h':
            print_usage();
            exit(0);
            break;
        case 'V':
            print_version();
            exit(0);
            break;
        case 'v':
            verbosity_level++;
            break;
        case 'X':
            xverbosity++;
            break;
        case 'n':
            numerals = 0;
            break;
        case 'x':
            numerals = -1;
            break;
        case 'c':
            auc = 1;
            auc_base = 753;
            break;
        case 'a':
            ad = 1;
            break;
        case 'd':
            if(optarg)
            {
                strncpy(datetype,optarg,16);
            }
            custom=1;
            break;
        case 'r':
            auc = 1;
            if(optarg)
                auc_base = atoi(optarg);
            break;
        case 'Y':
            if(optarg)
                p_ts->tm_year = atoi(optarg);
            break;
        case 'M':
            if(optarg)
                p_ts->tm_mon = atoi(optarg)-1; // 0-indexed months
            break;
        case 'D':
            if(optarg)
                p_ts->tm_mday = atoi(optarg);
            break;
        default:
            break;
        }
    }

    if(verbosity_level > 2)
        verbosity_level = 2;

    switch(verbosity_level) 
    {
    case 2:
        verbosity.year = 1;
    case 1:
        verbosity.day = 1;
        break;
    }

    if(optind < argc && custom == 1) {
        strncpy(dateval,argv[optind++],16);
    }

    if(custom)
        parse_date(datetype, dateval, p_ts);

    if(hodie_isleap(p_ts->tm_year)==1)
       leap=1;

    if(custom==0)
    {
	p_ts->tm_year=p_ts->tm_year+auc_base+1900;
    }
    
    p_ts->tm_mon++;


/* Setup the mode struct */
    m.numeral=1;
    
    to_roman(p_ts->tm_sec,sec,m);
    to_roman(p_ts->tm_min,min,m);
    to_roman(p_ts->tm_hour,hrs,m);
    to_roman(p_ts->tm_mday,day,m);
    to_roman(p_ts->tm_mon,mon,m);
    to_roman(p_ts->tm_year,year,m);

    date_to_roman(p_ts, verbose, verbosity, ad, leap);

    if(numerals!=0 && auc!=0)
    {
	printf("Hodie %s %s est\n",verbose,AUC[verbosity.day+1]);
    }
    else if(numerals!=0)
    {
	printf("Hodie %s est\n",verbose);
    }

    if(numerals==-1) /* The only case where both numerals and text exist */
	printf("\n");

    if(numerals < 1)
    {
	if(auc!=0)
	    printf("%s/%s-%s %s\n%s:%s:%s\n", 
		   day,mon,year,AUC[verbosity.day+1],hrs,min,sec);
	else
	    printf("%s/%s-%s\n%s:%s:%s\n", 
		   day,mon,year,hrs,min,sec);
    }	
    return 0;
}
